### <font color='purple'>  I decided to design and create a simple databse for MySQL. Here are the tables I created:</font>
    
<font color='purple'> **Book**</font>
<br>

<font color='purple'> **Author**</font>
<br>  
<font color='purple'> **book_auther association table**</font>
<br>  
<font color='purple'> **Unique genre**</font>
<br>  
<font color='purple'> **book_genre association table**</font>
<br>  

<font color='purple'> My goal was to create more tables, but cleaning the data in a way that I could import it to MySQL Workbench turned out to be a significant hassle.</font>



In [1]:
import pandas as pd
import numpy as np

import ast

import datetime
import warnings

warnings.filterwarnings('ignore')
warnings.filterwarnings("ignore", message="numpy.ufunc size changed")


import random


In [2]:
books_enriched= pd.read_csv( 'books_enriched_new_dataset.csv' )
books_enriched.head()

Unnamed: 0  index                            authors  average_rating  \
0           0      0                ['Suzanne Collins']            4.34   
1           1      1  ['J.K. Rowling', 'Mary GrandPré']            4.44   
2           2      2                ['Stephenie Meyer']            3.57   
3           3      3                     ['Harper Lee']            4.25   
4           4      4            ['F. Scott Fitzgerald']            3.89   

   best_book_id  book_id  books_count  \
0       2767052        1          272   
1             3        2          491   
2         41865        3          226   
3          2657        4          487   
4          4671        5         1356   

                                         description  \
0  WINNING MEANS FAME AND FORTUNE.LOSING MEANS CE...   
1  Harry Potter's life is miserable. His parents ...   
2  About three things I was absolutely positive.\...   
3  The unforgettable novel of a childhood in a sl...   
4  Alternate Cover Edition ISBN: 0743273567 (ISBN...   

                                              genres  goodreads_book_id  ...  \
0  ['young-adult', 'fiction', 'fantasy', 'science...            2767052  ...   
1  ['fantasy', 'fiction', 'young-adult', 'classics']                  3  ...   
2  ['young-adult', 'fantasy', 'romance', 'fiction...              41865  ...   
3  ['classics', 'fiction', 'historical-fiction', ...               2657  ...   
4  ['classics', 'fiction', 'historical-fiction', ...               4671  ...   

  ratings_3 ratings_4  ratings_5 ratings_count  \
0    560092   1481305    2706317       4780653   
1    455024   1156318    3011543       4602479   
2    793319    875073    1355439       3866839   
3    446835   1001952    1714267       3198671   
4    606158    936012     947718       2683664   

                                     small_image_url  \
0  https://images.gr-assets.com/books/1447303603s...   
1  https://images.gr-assets.com/books/1474154022s...   
2  https://images.gr-assets.com/books/1361039443s...   
3  https://images.gr-assets.com/books/1361975680s...   
4  https://images.gr-assets.com/books/1490528560s...   

                                               title  work_id  \
0            The Hunger Games (The Hunger Games, #1)  2792775   
1  Harry Potter and the Sorcerer's Stone (Harry P...  4640799   
2                            Twilight (Twilight, #1)  3212258   
3                              To Kill a Mockingbird  3275794   
4                                   The Great Gatsby   245494   

  work_ratings_count  work_text_reviews_count  \
0            4942365                   155254   
1            4800065                    75867   
2            3916824                    95009   
3            3340896                    72586   
4            2773745                    51992   

                           authors_2  
0                ['Suzanne Collins']  
1  ['J.K. Rowling', 'Mary GrandPré']  
2                ['Stephenie Meyer']  
3                     ['Harper Lee']  
4            ['F. Scott Fitzgerald']  

[5 rows x 30 columns]

In [3]:
books_enriched.shape

(10000, 30)

In [4]:
books_enriched.dtypes

Unnamed: 0                     int64
index                          int64
authors                       object
average_rating               float64
best_book_id                   int64
book_id                        int64
books_count                    int64
description                   object
genres                        object
goodreads_book_id              int64
image_url                     object
isbn                          object
isbn13                       float64
language_code                 object
original_publication_year    float64
original_title                object
pages                        float64
publishDate                   object
ratings_1                      int64
ratings_2                      int64
ratings_3                      int64
ratings_4                      int64
ratings_5                      int64
ratings_count                  int64
small_image_url               object
title                         object
work_id                        int64
w

In [5]:
books_enriched['original_title'].unique()

array(['The Hunger Games', "Harry Potter and the Philosopher's Stone",
       'Twilight', ..., "Tom Clancy's Net Force: Breaking Point",
       'Chasing The Night', 'The First World War'], dtype=object)

In [6]:
books_enriched['title'].unique()

array(['The Hunger Games (The Hunger Games, #1)',
       "Harry Potter and the Sorcerer's Stone (Harry Potter, #1)",
       'Twilight (Twilight, #1)', ...,
       "Breaking Point (Tom Clancy's Net Force, #4)",
       'Chasing The Night (Eve Duncan, #11; Catherine Ling, #1)',
       'The First World War'], dtype=object)

# <font color='purple'> Cleaning the dataset</font>

In [7]:
#Dropping some unnecessary columns

books_enriched = books_enriched.drop(['Unnamed: 0'], axis=1)
books_enriched.head()

index                            authors  average_rating  best_book_id  \
0      0                ['Suzanne Collins']            4.34       2767052   
1      1  ['J.K. Rowling', 'Mary GrandPré']            4.44             3   
2      2                ['Stephenie Meyer']            3.57         41865   
3      3                     ['Harper Lee']            4.25          2657   
4      4            ['F. Scott Fitzgerald']            3.89          4671   

   book_id  books_count                                        description  \
0        1          272  WINNING MEANS FAME AND FORTUNE.LOSING MEANS CE...   
1        2          491  Harry Potter's life is miserable. His parents ...   
2        3          226  About three things I was absolutely positive.\...   
3        4          487  The unforgettable novel of a childhood in a sl...   
4        5         1356  Alternate Cover Edition ISBN: 0743273567 (ISBN...   

                                              genres  goodreads_book_id  \
0  ['young-adult', 'fiction', 'fantasy', 'science...            2767052   
1  ['fantasy', 'fiction', 'young-adult', 'classics']                  3   
2  ['young-adult', 'fantasy', 'romance', 'fiction...              41865   
3  ['classics', 'fiction', 'historical-fiction', ...               2657   
4  ['classics', 'fiction', 'historical-fiction', ...               4671   

                                           image_url  ... ratings_3  \
0  https://images.gr-assets.com/books/1447303603m...  ...    560092   
1  https://images.gr-assets.com/books/1474154022m...  ...    455024   
2  https://images.gr-assets.com/books/1361039443m...  ...    793319   
3  https://images.gr-assets.com/books/1361975680m...  ...    446835   
4  https://images.gr-assets.com/books/1490528560m...  ...    606158   

   ratings_4 ratings_5  ratings_count  \
0    1481305   2706317        4780653   
1    1156318   3011543        4602479   
2     875073   1355439        3866839   
3    1001952   1714267        3198671   
4     936012    947718        2683664   

                                     small_image_url  \
0  https://images.gr-assets.com/books/1447303603s...   
1  https://images.gr-assets.com/books/1474154022s...   
2  https://images.gr-assets.com/books/1361039443s...   
3  https://images.gr-assets.com/books/1361975680s...   
4  https://images.gr-assets.com/books/1490528560s...   

                                               title  work_id  \
0            The Hunger Games (The Hunger Games, #1)  2792775   
1  Harry Potter and the Sorcerer's Stone (Harry P...  4640799   
2                            Twilight (Twilight, #1)  3212258   
3                              To Kill a Mockingbird  3275794   
4                                   The Great Gatsby   245494   

   work_ratings_count  work_text_reviews_count  \
0             4942365                   155254   
1             4800065                    75867   
2             3916824                    95009   
3             3340896                    72586   
4             2773745                    51992   

                           authors_2  
0                ['Suzanne Collins']  
1  ['J.K. Rowling', 'Mary GrandPré']  
2                ['Stephenie Meyer']  
3                     ['Harper Lee']  
4            ['F. Scott Fitzgerald']  

[5 rows x 29 columns]

### <font color='purple'> Checking the duplicated rows</font>

In [8]:
books_enriched.duplicated().sum()

0

### <font color='purple'> Checking the missing values, and filling or droping them</font>

In [9]:
#checking the missing values

print("\nNumber of missing values in each column: ")
print(books_enriched.isnull().sum())


Number of missing values in each column: 
index                          0
authors                        0
average_rating                 0
best_book_id                   0
book_id                        0
books_count                    0
description                   57
genres                         0
goodreads_book_id              0
image_url                      0
isbn                         700
isbn13                       585
language_code                  0
original_publication_year     21
original_title               585
pages                         73
publishDate                    8
ratings_1                      0
ratings_2                      0
ratings_3                      0
ratings_4                      0
ratings_5                      0
ratings_count                  0
small_image_url                0
title                          0
work_id                        0
work_ratings_count             0
work_text_reviews_count        0
authors_2                      0


#### <font color='green'> We have 2 title columns here "original_title" and "title", data in the "original_title" is way cleaner than data in the "title" column but there are 585 missing values in "original_title" and no missing values in "title" so I want to fill the missing cells in "original_title" with corresponding values from the 'title' column.</font>

In [10]:
books_enriched['original_title'].fillna(books_enriched['title'], inplace=True)

#### <font color='green'> I want to check the values in publishDate and original_publication_year and then possibly drop the missing values. I tried to change the type for publishDate to datetime and extrat the year and then fillna the missing values in  original_publication_year but it was not easy so I prefer to drop these rows as the amount is not significant</font>

In [11]:
print(books_enriched['publishDate'].unique())

['09/14/08' '11/01/03' '09/06/06' ... "('6', '7', '2005')"
 "('10', '19', '2010')" "('5', '16', '2000')"]


In [12]:
print(books_enriched['original_publication_year'].unique())

[ 2008.  1997.  2005.  1960.  1925.  2012.  1937.  1951.  2000.  1813.
  2003.  2011.  1947.  2009.  1999.  2010.  2002.  1998.  2007.  1954.
  1595.  1993.  1950.  1996.  2006.  1868.  1847.  2001.  1974.  1979.
  1932.  1884.  1952.  2015.  1995.  2004.  1969.  1936.  1985.  1977.
  1811.  -720.  1946.  1859.  1964.  1989.  1958.  1973.  1967.  1911.
  1891.  1897.  1963.  1844.  2013.  1862.  1961.  1876.  1955.  1991.
  1600.  1962.  1939.  1908.  1850.  2014.  1606.  1860.  1942.  1978.
  1815.  1877.  1986.  1866.  1922.  1965.  1988.  1987.  1851.  1843.
  1976.  1994.  1915.  1956.    nan  1980.  1865.  1817.  1982.  1957.
  1926.  1943.  1938.  1981.  2016.  1992.  1990.  1984.  1972.  1882.
  1895.  1899.  1983.  1953.  -750.  1900.  1975.  1971.  1603.  1929.
  1838.  1903.  1886.  1940.  -500.  1887.  1931.  1611.  1814.  1719.
  1513.  1880.  1923.  1869.  1968.  1849.  1892.  1904.  1726.  1598.
   975.  1905.  1935.  1948.  1856.  1759.  1959.  1605.  1901.  1970.
  1902

In [13]:
#I also see negative numbers in 'original_publication_year' so let's see how many they are

num_negative_values = (books_enriched['original_publication_year'] < 0).sum()
num_negative_values

31

In [14]:
#I want to see what are the rows with negative values in 'original_publication_year' and if I can get the publish year from other sources and fill nans

books_enriched[books_enriched['original_publication_year'] < 0]

index                                            authors  \
71       71  ['Homer', 'Robert Fagles', 'E.V. Rieu', 'Frédé...   
311     311  ['Homer', 'Robert Fagles', 'Frédéric Mugler', ...   
368     368                       ['Sun Tzu', 'Thomas Cleary']   
722     722                           ['Plato', 'Desmond Lee']   
773     773                       ['Sophocles', 'J.E. Thomas']   
1025   1025  ['Lao Tzu', 'Gia-Fu Feng', 'Jane English', 'Ch...   
1044   1044           ['Aesop', 'Laura Harris', 'Laura Gibbs']   
1199   1199                    ['Virgil', 'Robert Fitzgerald']   
1429   1429                       ['Sophocles', 'J.E. Thomas']   
1924   1924  ['Sophocles', 'Dudley Fitts', 'Elena Bono', 'R...   
1946   1946                      ['Anonymous', 'N.K. Sandars']   
2009   2009         ['Homer', 'Robert Fagles', 'Bernard Knox']   
2187   2187  ['Krishna-Dwaipayana Vyasa', 'Simon Brodbeck',...   
2577   2577                        ['Euripides', 'Rex Warner']   
3157   3157  ['Herodotus', 'Aubrey de Sélincourt', 'John M....   
3558   3558                      ['Plato', 'Christopher Gill']   
3611   3611  ['Aristotle', 'J.A.K. Thomson', 'Jonathan Barn...   
3626   3626                                      ['Aristotle']   
3795   3795  ['Aeschylus', 'Robert Fagles', 'William Bedell...   
3947   3947        ['Plato', 'G.M.A. Grube', 'John M. Cooper']   
3949   3949    ['Thucydides', 'Rex Warner', 'Moses I. Finley']   
4112   4112                    ['Aristophanes', 'Sarah Ruden']   
4393   4393                         ['Plato', 'James J. Helm']   
5022   5022  ['Anonymous', 'Ananda Maitreya', 'Thich Nhat H...   
5439   5439  ['Anonymous', 'Richard Wilhelm', 'Cary F. Bayn...   
6006   6006        ['Plato', 'G.M.A. Grube', 'John M. Cooper']   
6651   6651                          ['Confucius', 'D.C. Lau']   
7159   7159                         ['Aristotle', 'Joe Sachs']   
7684   7684  ['Aeschylus', 'Judith Affleck', 'Philip de May...   
7700   7700                     ['Aristotle', 'Malcolm Heath']   
7950   7950                      ['Anonymous', 'Juan Mascaró']   

      average_rating  best_book_id  book_id  books_count  \
71              3.73          1381       79         1703   
311             3.83          1371      341         1726   
368             3.95         10534      403         1865   
722             3.90         30289      772          842   
773             3.67          1554      824          498   
1025            4.31         67896     1099         1265   
1044            4.05         21348     1120          942   
1199            3.81         12914     1280          830   
1429            3.60          7728     1521          423   
1924            3.96          1540     2052          331   
1946            3.63         19351     2076          266   
2009            4.03          1375     2142          255   
2187            4.10         99944     2336          596   
2577            3.83        752900     2763          304   
3157            3.97          1362     3404          262   
3558            4.01         81779     3868          458   
3611            3.91         19068     3930          447   
3626            3.91         19083     3949          355   
3795            3.99          1519     4149          255   
3947            4.06         22632     4330          102   
3949            3.87        261243     4332          336   
4112            3.84          1591     4537          222   
4393            4.15         73945     4880          306   
5022            4.29        159964     5637          259   
5439            4.18        534289     6166          140   
6006            4.10         30292     6896           25   
6651            3.82         27297     7778          348   
7159            4.00        208036     8472          201   
7684            3.82          1524     9281          190   
7700            3.81         13270     9304          456   
7950            4.20        29088

#### <font color='green'>I checked a few titles, I can't find an accurate publish year for them so I'm gonna drop the rows with missing and negative values for 'original_publication_year'.</font >

In [15]:
# Drop rows where 'original_publication_year' is NaN or negative

books_enriched = books_enriched.dropna(subset=['original_publication_year'])
books_enriched = books_enriched[books_enriched['original_publication_year'] >= 0]

# Reset the index
books_enriched.reset_index(drop=True, inplace=True)

books_enriched.head()

index                            authors  average_rating  best_book_id  \
0      0                ['Suzanne Collins']            4.34       2767052   
1      1  ['J.K. Rowling', 'Mary GrandPré']            4.44             3   
2      2                ['Stephenie Meyer']            3.57         41865   
3      3                     ['Harper Lee']            4.25          2657   
4      4            ['F. Scott Fitzgerald']            3.89          4671   

   book_id  books_count                                        description  \
0        1          272  WINNING MEANS FAME AND FORTUNE.LOSING MEANS CE...   
1        2          491  Harry Potter's life is miserable. His parents ...   
2        3          226  About three things I was absolutely positive.\...   
3        4          487  The unforgettable novel of a childhood in a sl...   
4        5         1356  Alternate Cover Edition ISBN: 0743273567 (ISBN...   

                                              genres  goodreads_book_id  \
0  ['young-adult', 'fiction', 'fantasy', 'science...            2767052   
1  ['fantasy', 'fiction', 'young-adult', 'classics']                  3   
2  ['young-adult', 'fantasy', 'romance', 'fiction...              41865   
3  ['classics', 'fiction', 'historical-fiction', ...               2657   
4  ['classics', 'fiction', 'historical-fiction', ...               4671   

                                           image_url  ... ratings_3  \
0  https://images.gr-assets.com/books/1447303603m...  ...    560092   
1  https://images.gr-assets.com/books/1474154022m...  ...    455024   
2  https://images.gr-assets.com/books/1361039443m...  ...    793319   
3  https://images.gr-assets.com/books/1361975680m...  ...    446835   
4  https://images.gr-assets.com/books/1490528560m...  ...    606158   

   ratings_4 ratings_5  ratings_count  \
0    1481305   2706317        4780653   
1    1156318   3011543        4602479   
2     875073   1355439        3866839   
3    1001952   1714267        3198671   
4     936012    947718        2683664   

                                     small_image_url  \
0  https://images.gr-assets.com/books/1447303603s...   
1  https://images.gr-assets.com/books/1474154022s...   
2  https://images.gr-assets.com/books/1361039443s...   
3  https://images.gr-assets.com/books/1361975680s...   
4  https://images.gr-assets.com/books/1490528560s...   

                                               title  work_id  \
0            The Hunger Games (The Hunger Games, #1)  2792775   
1  Harry Potter and the Sorcerer's Stone (Harry P...  4640799   
2                            Twilight (Twilight, #1)  3212258   
3                              To Kill a Mockingbird  3275794   
4                                   The Great Gatsby   245494   

   work_ratings_count  work_text_reviews_count  \
0             4942365                   155254   
1             4800065                    75867   
2             3916824                    95009   
3             3340896                    72586   
4             2773745                    51992   

                           authors_2  
0                ['Suzanne Collins']  
1  ['J.K. Rowling', 'Mary GrandPré']  
2                ['Stephenie Meyer']  
3                     ['Harper Lee']  
4            ['F. Scott Fitzgerald']  

[5 rows x 29 columns]

In [16]:
books_enriched.shape

(9948, 29)

In [17]:
print("\nNumber of missing values in each column: ")
print(books_enriched.isnull().sum())


Number of missing values in each column: 
index                          0
authors                        0
average_rating                 0
best_book_id                   0
book_id                        0
books_count                    0
description                   54
genres                         0
goodreads_book_id              0
image_url                      0
isbn                         697
isbn13                       582
language_code                  0
original_publication_year      0
original_title                 0
pages                         71
publishDate                    8
ratings_1                      0
ratings_2                      0
ratings_3                      0
ratings_4                      0
ratings_5                      0
ratings_count                  0
small_image_url                0
title                          0
work_id                        0
work_ratings_count             0
work_text_reviews_count        0
authors_2                      0


### <font color='purple'>Droping unnessary columns </font>

**Work IDs** are unique identification numbers that represent books in the Goodreads catalog. You can find the Work ID of a book by clicking on All Editions in the book detail page.


All the ones prefixed with "work_" mean stats on all editions (not just the isbn passed in).

id is the unique goodreads identifier for that isbn.


In [18]:
books_enriched_cleaned = books_enriched.drop(['authors_2', 'title' ], axis=1)
books_enriched_cleaned.head()

index                            authors  average_rating  best_book_id  \
0      0                ['Suzanne Collins']            4.34       2767052   
1      1  ['J.K. Rowling', 'Mary GrandPré']            4.44             3   
2      2                ['Stephenie Meyer']            3.57         41865   
3      3                     ['Harper Lee']            4.25          2657   
4      4            ['F. Scott Fitzgerald']            3.89          4671   

   book_id  books_count                                        description  \
0        1          272  WINNING MEANS FAME AND FORTUNE.LOSING MEANS CE...   
1        2          491  Harry Potter's life is miserable. His parents ...   
2        3          226  About three things I was absolutely positive.\...   
3        4          487  The unforgettable novel of a childhood in a sl...   
4        5         1356  Alternate Cover Edition ISBN: 0743273567 (ISBN...   

                                              genres  goodreads_book_id  \
0  ['young-adult', 'fiction', 'fantasy', 'science...            2767052   
1  ['fantasy', 'fiction', 'young-adult', 'classics']                  3   
2  ['young-adult', 'fantasy', 'romance', 'fiction...              41865   
3  ['classics', 'fiction', 'historical-fiction', ...               2657   
4  ['classics', 'fiction', 'historical-fiction', ...               4671   

                                           image_url  ... ratings_1  \
0  https://images.gr-assets.com/books/1447303603m...  ...     66715   
1  https://images.gr-assets.com/books/1474154022m...  ...     75504   
2  https://images.gr-assets.com/books/1361039443m...  ...    456191   
3  https://images.gr-assets.com/books/1361975680m...  ...     60427   
4  https://images.gr-assets.com/books/1490528560m...  ...     86236   

   ratings_2 ratings_3  ratings_4 ratings_5  ratings_count  \
0     127936    560092    1481305   2706317        4780653   
1     101676    455024    1156318   3011543        4602479   
2     436802    793319     875073   1355439        3866839   
3     117415    446835    1001952   1714267        3198671   
4     197621    606158     936012    947718        2683664   

                                     small_image_url  work_id  \
0  https://images.gr-assets.com/books/1447303603s...  2792775   
1  https://images.gr-assets.com/books/1474154022s...  4640799   
2  https://images.gr-assets.com/books/1361039443s...  3212258   
3  https://images.gr-assets.com/books/1361975680s...  3275794   
4  https://images.gr-assets.com/books/1490528560s...   245494   

   work_ratings_count  work_text_reviews_count  
0             4942365                   155254  
1             4800065                    75867  
2             3916824                    95009  
3             3340896                    72586  
4             2773745                    51992  

[5 rows x 27 columns]

books_enriched_cleaned = books_enriched.drop(['authors_2', 'best_book_id','image_url', 'small_image_url', 'work_id', 'book_id', 'isbn','isbn13','publishDate', 'title' ], axis=1)
books_enriched_cleaned.head()

In [19]:
print("\nNumber of missing values in each column: ")
print(books_enriched_cleaned.isnull().sum())


Number of missing values in each column: 
index                          0
authors                        0
average_rating                 0
best_book_id                   0
book_id                        0
books_count                    0
description                   54
genres                         0
goodreads_book_id              0
image_url                      0
isbn                         697
isbn13                       582
language_code                  0
original_publication_year      0
original_title                 0
pages                         71
publishDate                    8
ratings_1                      0
ratings_2                      0
ratings_3                      0
ratings_4                      0
ratings_5                      0
ratings_count                  0
small_image_url                0
work_id                        0
work_ratings_count             0
work_text_reviews_count        0
dtype: int64


#### <font color='green'> I want to fillna the missing values for "pages" with average number of pages and I fillna the missigng values of description with the string "No description".</font>

In [20]:
# Calculate the average number of pages
avg_pages = books_enriched_cleaned['pages'].mean()
avg_pages

359.72441024602614

In [21]:
books_enriched_cleaned['pages'].fillna(avg_pages, inplace=True)

In [22]:
books_enriched_cleaned['description'].fillna('No description', inplace=True)

In [23]:
print("\nNumber of missing values in each column: ")
print(books_enriched_cleaned.isnull().sum())


Number of missing values in each column: 
index                          0
authors                        0
average_rating                 0
best_book_id                   0
book_id                        0
books_count                    0
description                    0
genres                         0
goodreads_book_id              0
image_url                      0
isbn                         697
isbn13                       582
language_code                  0
original_publication_year      0
original_title                 0
pages                          0
publishDate                    8
ratings_1                      0
ratings_2                      0
ratings_3                      0
ratings_4                      0
ratings_5                      0
ratings_count                  0
small_image_url                0
work_id                        0
work_ratings_count             0
work_text_reviews_count        0
dtype: int64


#### <font color='green'> No More Missing Values : ) at least at this stage.</font>

In [24]:
books_enriched_cleaned.shape

(9948, 27)

### <font color='purple'> Cleaning the values in the authors column</font>

In [25]:
books_enriched_cleaned['authors'].value_counts()

['Stephen King']                      58
['Nora Roberts']                      46
['Terry Pratchett']                   40
['Agatha Christie']                   38
['Dean Koontz']                       34
                                      ..
['Bob Ong']                            1
['Anne Tyler', 'Jennifer Bassett']     1
['Loung Ung']                          1
['Betty Mahmoody']                     1
['[John Keegan]']                      1
Name: authors, Length: 5017, dtype: int64

To remove the square brackets and quotes and convert the values to plain strings, I use the literal_eval function from the ast module in Python to convert the string to a list, and then use the join method to convert the list back to a string.


In [26]:
# Convert the string to a list and join the list elements to a string
books_enriched_cleaned['authors'] = books_enriched_cleaned['authors'].apply(lambda x: ', '.join(ast.literal_eval(x)))
books_enriched_cleaned.head()

index                      authors  average_rating  best_book_id  book_id  \
0      0              Suzanne Collins            4.34       2767052        1   
1      1  J.K. Rowling, Mary GrandPré            4.44             3        2   
2      2              Stephenie Meyer            3.57         41865        3   
3      3                   Harper Lee            4.25          2657        4   
4      4          F. Scott Fitzgerald            3.89          4671        5   

   books_count                                        description  \
0          272  WINNING MEANS FAME AND FORTUNE.LOSING MEANS CE...   
1          491  Harry Potter's life is miserable. His parents ...   
2          226  About three things I was absolutely positive.\...   
3          487  The unforgettable novel of a childhood in a sl...   
4         1356  Alternate Cover Edition ISBN: 0743273567 (ISBN...   

                                              genres  goodreads_book_id  \
0  ['young-adult', 'fiction', 'fantasy', 'science...            2767052   
1  ['fantasy', 'fiction', 'young-adult', 'classics']                  3   
2  ['young-adult', 'fantasy', 'romance', 'fiction...              41865   
3  ['classics', 'fiction', 'historical-fiction', ...               2657   
4  ['classics', 'fiction', 'historical-fiction', ...               4671   

                                           image_url  ... ratings_1  \
0  https://images.gr-assets.com/books/1447303603m...  ...     66715   
1  https://images.gr-assets.com/books/1474154022m...  ...     75504   
2  https://images.gr-assets.com/books/1361039443m...  ...    456191   
3  https://images.gr-assets.com/books/1361975680m...  ...     60427   
4  https://images.gr-assets.com/books/1490528560m...  ...     86236   

   ratings_2 ratings_3  ratings_4 ratings_5  ratings_count  \
0     127936    560092    1481305   2706317        4780653   
1     101676    455024    1156318   3011543        4602479   
2     436802    793319     875073   1355439        3866839   
3     117415    446835    1001952   1714267        3198671   
4     197621    606158     936012    947718        2683664   

                                     small_image_url  work_id  \
0  https://images.gr-assets.com/books/1447303603s...  2792775   
1  https://images.gr-assets.com/books/1474154022s...  4640799   
2  https://images.gr-assets.com/books/1361039443s...  3212258   
3  https://images.gr-assets.com/books/1361975680s...  3275794   
4  https://images.gr-assets.com/books/1490528560s...   245494   

   work_ratings_count  work_text_reviews_count  
0             4942365                   155254  
1             4800065                    75867  
2             3916824                    95009  
3             3340896                    72586  
4             2773745                    51992  

[5 rows x 27 columns]

In [27]:
books_enriched_cleaned['genres'].value_counts()

['fantasy', 'fiction']                                                     151
['fiction', 'contemporary']                                                133
['fantasy', 'young-adult', 'fiction']                                      132
['science-fiction', 'fiction', 'fantasy']                                   80
['fantasy', 'fiction', 'young-adult']                                       77
                                                                          ... 
['fiction', 'mystery', 'suspense', 'chick-lit', 'contemporary']              1
['fantasy', 'horror', 'fiction', 'science-fiction', 'thriller']              1
['mystery', 'paranormal', 'fantasy', 'fiction', 'romance', 'chick-lit']      1
['fantasy', 'fiction', 'art']                                                1
['history', 'nonfiction', 'historical-fiction']                              1
Name: genres, Length: 4327, dtype: int64

In [28]:

#keeping only the official goodreads genres from the genres column

official_genres = {"art", "biography", "business", "chick-lit", "children's", "christian", "classics",
          "comics", "contemporary", "cookbooks", "crime", "books", "fantasy", "fiction",
          "gay-and-lesbian", "graphic-novels", "historical-fiction", "history", "horror",
          "humor-and-comedy", "manga", "memoir", "music", "mystery", "nonfiction", "paranormal",
          "philosophy", "poetry", "psychology", "religion", "romance", "science", "science-fiction", 
          "self-help", "suspense", "spirituality", "sports", "thriller", "travel", "young-adult"}

# Convert the values of the genre columns that are string to a list and keep only the elements that are in official_genres
books_enriched_cleaned['genres'] = books_enriched_cleaned['genres'].apply(lambda x: [element for element in ast.literal_eval(x) if element in official_genres])

In [29]:
books_enriched_cleaned.head()

index                      authors  average_rating  best_book_id  book_id  \
0      0              Suzanne Collins            4.34       2767052        1   
1      1  J.K. Rowling, Mary GrandPré            4.44             3        2   
2      2              Stephenie Meyer            3.57         41865        3   
3      3                   Harper Lee            4.25          2657        4   
4      4          F. Scott Fitzgerald            3.89          4671        5   

   books_count                                        description  \
0          272  WINNING MEANS FAME AND FORTUNE.LOSING MEANS CE...   
1          491  Harry Potter's life is miserable. His parents ...   
2          226  About three things I was absolutely positive.\...   
3          487  The unforgettable novel of a childhood in a sl...   
4         1356  Alternate Cover Edition ISBN: 0743273567 (ISBN...   

                                              genres  goodreads_book_id  \
0  [young-adult, fiction, fantasy, science-fictio...            2767052   
1          [fantasy, fiction, young-adult, classics]                  3   
2  [young-adult, fantasy, romance, fiction, paran...              41865   
3  [classics, fiction, historical-fiction, young-...               2657   
4   [classics, fiction, historical-fiction, romance]               4671   

                                           image_url  ... ratings_1  \
0  https://images.gr-assets.com/books/1447303603m...  ...     66715   
1  https://images.gr-assets.com/books/1474154022m...  ...     75504   
2  https://images.gr-assets.com/books/1361039443m...  ...    456191   
3  https://images.gr-assets.com/books/1361975680m...  ...     60427   
4  https://images.gr-assets.com/books/1490528560m...  ...     86236   

   ratings_2 ratings_3  ratings_4 ratings_5  ratings_count  \
0     127936    560092    1481305   2706317        4780653   
1     101676    455024    1156318   3011543        4602479   
2     436802    793319     875073   1355439        3866839   
3     117415    446835    1001952   1714267        3198671   
4     197621    606158     936012    947718        2683664   

                                     small_image_url  work_id  \
0  https://images.gr-assets.com/books/1447303603s...  2792775   
1  https://images.gr-assets.com/books/1474154022s...  4640799   
2  https://images.gr-assets.com/books/1361039443s...  3212258   
3  https://images.gr-assets.com/books/1361975680s...  3275794   
4  https://images.gr-assets.com/books/1490528560s...   245494   

   work_ratings_count  work_text_reviews_count  
0             4942365                   155254  
1             4800065                    75867  
2             3916824                    95009  
3             3340896                    72586  
4             2773745                    51992  

[5 rows x 27 columns]

In [30]:
row_index = 0  
print(books_enriched_cleaned.loc[row_index, 'genres'])

['young-adult', 'fiction', 'fantasy', 'science-fiction', 'romance']


In [31]:
books_enriched_cleaned.shape

(9948, 27)

In [32]:
books_enriched_cleaned = books_enriched_cleaned.drop(['isbn' ], axis=1)


In [33]:
books_enriched_cleaned.head()

index                      authors  average_rating  best_book_id  book_id  \
0      0              Suzanne Collins            4.34       2767052        1   
1      1  J.K. Rowling, Mary GrandPré            4.44             3        2   
2      2              Stephenie Meyer            3.57         41865        3   
3      3                   Harper Lee            4.25          2657        4   
4      4          F. Scott Fitzgerald            3.89          4671        5   

   books_count                                        description  \
0          272  WINNING MEANS FAME AND FORTUNE.LOSING MEANS CE...   
1          491  Harry Potter's life is miserable. His parents ...   
2          226  About three things I was absolutely positive.\...   
3          487  The unforgettable novel of a childhood in a sl...   
4         1356  Alternate Cover Edition ISBN: 0743273567 (ISBN...   

                                              genres  goodreads_book_id  \
0  [young-adult, fiction, fantasy, science-fictio...            2767052   
1          [fantasy, fiction, young-adult, classics]                  3   
2  [young-adult, fantasy, romance, fiction, paran...              41865   
3  [classics, fiction, historical-fiction, young-...               2657   
4   [classics, fiction, historical-fiction, romance]               4671   

                                           image_url  ...  ratings_1  \
0  https://images.gr-assets.com/books/1447303603m...  ...      66715   
1  https://images.gr-assets.com/books/1474154022m...  ...      75504   
2  https://images.gr-assets.com/books/1361039443m...  ...     456191   
3  https://images.gr-assets.com/books/1361975680m...  ...      60427   
4  https://images.gr-assets.com/books/1490528560m...  ...      86236   

  ratings_2  ratings_3 ratings_4  ratings_5 ratings_count  \
0    127936     560092   1481305    2706317       4780653   
1    101676     455024   1156318    3011543       4602479   
2    436802     793319    875073    1355439       3866839   
3    117415     446835   1001952    1714267       3198671   
4    197621     606158    936012     947718       2683664   

                                     small_image_url  work_id  \
0  https://images.gr-assets.com/books/1447303603s...  2792775   
1  https://images.gr-assets.com/books/1474154022s...  4640799   
2  https://images.gr-assets.com/books/1361039443s...  3212258   
3  https://images.gr-assets.com/books/1361975680s...  3275794   
4  https://images.gr-assets.com/books/1490528560s...   245494   

   work_ratings_count  work_text_reviews_count  
0             4942365                   155254  
1             4800065                    75867  
2             3916824                    95009  
3             3340896                    72586  
4             2773745                    51992  

[5 rows x 26 columns]

In [34]:
books_enriched_cleaned.shape

(9948, 26)

In [35]:
#Seems like I still have lists:
    
# Function to check if I have other lists
def contains_list(element):
    return isinstance(element, list)

# Applying the function to each cell to identify columns with lists
list_columns = books_enriched_cleaned.applymap(contains_list)

# Checking which columns have lists by summing the True values 
columns_with_lists = list_columns.sum() > 0

# Convert the columns with lists to tuples
for col in columns_with_lists[columns_with_lists].index:
    books_enriched_cleaned[col] = books_enriched_cleaned[col].apply(tuple)

# Identifying the duplicate rows
duplicate_rows = books_enriched_cleaned.duplicated()

# Filtering the duplicate_rows with boolean array
duplicate_rows_df = books_enriched_cleaned[duplicate_rows]

# Dropping the dups
books_enriched_cleaned_unique = books_enriched_cleaned.drop_duplicates()


In [36]:
books_enriched_cleaned.duplicated().sum()

0

In [37]:
books_enriched_cleaned.columns

Index(['index', 'authors', 'average_rating', 'best_book_id', 'book_id',
       'books_count', 'description', 'genres', 'goodreads_book_id',
       'image_url', 'isbn13', 'language_code', 'original_publication_year',
       'original_title', 'pages', 'publishDate', 'ratings_1', 'ratings_2',
       'ratings_3', 'ratings_4', 'ratings_5', 'ratings_count',
       'small_image_url', 'work_id', 'work_ratings_count',
       'work_text_reviews_count'],
      dtype='object')

### <font color='purple'> Breaking down the books_enriched_cleaned table to multiplr tables for my MYSQL databse</font>

#### <font color='purple'>Book table: goodreads_book_id, isbn13, title, description,pages, language_code, original_publication_year </font>

In [38]:
Book=books_enriched_cleaned[['goodreads_book_id', 'original_title', 'description', 'pages', 'language_code', 'original_publication_year']]
Book.head()

goodreads_book_id                            original_title  \
0            2767052                          The Hunger Games   
1                  3  Harry Potter and the Philosopher's Stone   
2              41865                                  Twilight   
3               2657                     To Kill a Mockingbird   
4               4671                          The Great Gatsby   

                                         description  pages language_code  \
0  WINNING MEANS FAME AND FORTUNE.LOSING MEANS CE...  374.0           eng   
1  Harry Potter's life is miserable. His parents ...  309.0           eng   
2  About three things I was absolutely positive.\...  501.0           eng   
3  The unforgettable novel of a childhood in a sl...  324.0           eng   
4  Alternate Cover Edition ISBN: 0743273567 (ISBN...  200.0           eng   

   original_publication_year  
0                     2008.0  
1                     1997.0  
2                     2005.0  
3                     1960.0  
4                     1925.0

In [39]:
Book['goodreads_book_id'].dtype

dtype('int64')

In [40]:
Book['original_title'].dtype

dtype('O')

In [41]:
Book['original_title'] = Book['original_title'].str.replace("'", "''") # Escape single quotes

Book['original_title']  = Book['original_title'] .str.replace("[^\w\s]", "") # Remove special characters

In [42]:
Book['original_title'] = Book['original_title'].str.strip() # Remove leading/trailing spaces

In [43]:
Book['original_title'] = Book['original_title'].astype('str')

Book['original_title'] = Book['original_title'].apply(lambda x: str(x) if pd.notnull(x) else x)

In [44]:
nan_count_original_title = Book['original_title'].isnull().sum()
nan_count_original_title

0

In [45]:
Book['original_title'].dtype

dtype('O')

In [46]:
Book.drop_duplicates(inplace=True)

In [47]:
Book['description'] = Book['description'].str.replace("'", "''") # Escape single quotes

Book['description'] = Book['description'] .str.replace("[^\w\s]", "") # Remove special characters

Book['description']= Book['description'].str.strip() # Remove leading/trailing spaces

In [48]:
Book['description'].dtype

dtype('O')

In [49]:
Book['description'] = Book['description'].astype('str')

Book['description'] = Book['description'].apply(lambda x: str(x) if pd.notnull(x) else x)

In [50]:
nan_count_description = Book['description'].isnull().sum()
nan_count_description

0

In [51]:
Book['pages'] = Book['pages'].astype(int)

In [52]:
Book['pages'].dtype

dtype('int64')

In [53]:
Book['pages'].unique

<bound method Series.unique of 0       374
1       309
2       501
3       324
4       200
       ... 
9943    816
9944    436
9945    368
9946    362
9947    475
Name: pages, Length: 9948, dtype: int64>

In [54]:
Book['language_code'].dtype

dtype('O')

In [55]:
# Escape single quotes
Book['language_code'] = Book['language_code'].str.replace("'", "''") 

# Remove special characters
Book['language_code']  = Book['language_code'] .str.replace("[^\w\s]", "") 

# Remove leading/trailing spaces
Book['language_code'] = Book['language_code'].str.strip() 

In [56]:
Book['language_code'] = Book['language_code'].astype('str')

Book['language_code'] = Book['language_code'].apply(lambda x: str(x) if pd.notnull(x) else x)

In [57]:
nan_count_original_title = Book['language_code'].isnull().sum()
nan_count_original_title

0

In [58]:
Book['original_publication_year'].dtype

dtype('float64')

In [59]:
Book['original_publication_year'] = Book['original_publication_year'].astype(int)

In [60]:
print(Book.isnull().sum())

goodreads_book_id            0
original_title               0
description                  0
pages                        0
language_code                0
original_publication_year    0
dtype: int64


In [61]:
Book = Book.drop(['description' ], axis=1)
Book.head()

goodreads_book_id                           original_title  pages  \
0            2767052                         The Hunger Games    374   
1                  3  Harry Potter and the Philosophers Stone    309   
2              41865                                 Twilight    501   
3               2657                    To Kill a Mockingbird    324   
4               4671                         The Great Gatsby    200   

  language_code  original_publication_year  
0           eng                       2008  
1           eng                       1997  
2           eng                       2005  
3           eng                       1960  
4           eng                       1925

In [62]:
Book.to_csv('CLEAN_Final_Book.csv', index=False)

In [63]:
CLEAN_Final_Book = pd.read_csv('CLEAN_Final_Book.csv', encoding='utf-8')
CLEAN_Final_Book.head()

goodreads_book_id                           original_title  pages  \
0            2767052                         The Hunger Games    374   
1                  3  Harry Potter and the Philosophers Stone    309   
2              41865                                 Twilight    501   
3               2657                    To Kill a Mockingbird    324   
4               4671                         The Great Gatsby    200   

  language_code  original_publication_year  
0           eng                       2008  
1           eng                       1997  
2           eng                       2005  
3           eng                       1960  
4           eng                       1925

In [64]:
problem_rows = CLEAN_Final_Book[CLEAN_Final_Book.apply(lambda row: row.astype(str).str.contains('Ã').any(), axis=1)]
problem_rows

Empty DataFrame
Columns: [goodreads_book_id, original_title, pages, language_code, original_publication_year]
Index: []

In [65]:
def contains_non_ascii(s):
    return any(ord(c) >= 128 for c in s)

# Find rows that contain non-ASCII characters
problem_rows = CLEAN_Final_Book[CLEAN_Final_Book.apply(lambda row: row.astype(str).apply(contains_non_ascii).any(), axis=1)]

print(problem_rows)

      goodreads_book_id             original_title  pages language_code  \
13              2429135      Män som hatar kvinnor    465           eng   
84                  320       Cien años de soledad    417           eng   
98                24280             Les Misérables   1463           eng   
125             6892870  Luftslottet som sprängdes    566           eng   
145               49552                  LÉtranger    123           eng   
...                 ...                        ...    ...           ...   
9818            6448482                  ソウルイーター 2    192           eng   
9847            2309720                     アラクレ 1    200           eng   
9860             484167                     千年の雪 1    209           eng   
9878            6081686                  会長はメイド様 2    208           eng   
9917            1087207                 桜蘭高校ホスト部 4    271           eng   

      original_publication_year  
13                         2005  
84                         1967

#### <font color='red'> The reason I could not import this book data to MYSQL workbench was non-ASCII characters in the original_title column (non-Latin scripts (like Cyrillic, Greek, or Japanese characters). I will drop them. </font>

In [66]:
# Function to remove non-ASCII characters from a string
def remove_non_ascii(s):
    if isinstance(s, str):
        return "".join(c for c in s if ord(c) < 128)
    else:
        return s

CLEAN_Final_Book['original_title'] = CLEAN_Final_Book['original_title'].apply(remove_non_ascii)

In [67]:
CLEAN_Final_Book.to_csv('CLEAN_Final_Book.csv', index=False)

Book = Book.drop(['original_publication_year'], axis=1)

import csv

def preprocess_csv(input_file, output_file, num_columns):
    with open(input_file, 'r') as read_obj, open(output_file, 'w') as write_obj:
        csv_reader = csv.reader(read_obj)
        csv_writer = csv.writer(write_obj)
        row = next(csv_reader)
        while row:
            while len(row) < num_columns:
                try:
                    row += next(csv_reader)
                except StopIteration:
                    row = None
                    break
            if row:
                csv_writer.writerow(row)
                try:
                    row = next(csv_reader)
                except StopIteration:
                    row = None

preprocess_csv('Final_Book_SQL.csv', 'CLEANED_Final_Book.csv', 6)

#### <font color='purple'> Author_book table:goodreads_book_id, title, </font>

In [68]:
#First I want to check if I have the same author for multiple books

# Group by 'authors' column and count the number of 'original_title'
book_counts_per_author = books_enriched_cleaned.groupby('authors')['original_title'].count()

# Filter for authors with more than one book
multiple_books_per_author = book_counts_per_author[book_counts_per_author > 1]

print(multiple_books_per_author)

authors
A. Meredith Walters              2
A.A. Milne, Ernest H. Shepard    6
A.G. Howard                      2
A.N. Roquelaure, Anne Rice       3
A.S. Byatt                       2
                                ..
غسان كنفاني                      2
محمد صادق                        3
مصطفى محمود                      2
يوسف زيدان                       2
يوسف زيدان, Youssef Ziedan       2
Name: original_title, Length: 1410, dtype: int64


In [69]:
counts = books_enriched_cleaned['authors'].value_counts()

author_name = 'A.S. Byatt'  
print(counts[author_name])

2


#### In some cases I have multiple author names in one row separated by a comma, I want to explode them to separated rows and have one author per row

In [70]:
# Split the 'authors' column into a list of authors
books_enriched_cleaned['authors'] = books_enriched_cleaned['authors'].str.split(',')

# Using explode to create one row per author
books_enriched_cleaned_authors = books_enriched_cleaned.explode('authors')

# strip leading/trailing whitespaces caused by splitting
books_enriched_cleaned_authors['authors'] = books_enriched_cleaned_authors['authors'].str.strip()

In [71]:
# Group by 'authors' column and count the number of 'original_title'
book_counts_per_author_2 = books_enriched_cleaned_authors.groupby('authors')['original_title'].count()

# Filter for authors with more than one book
multiple_books_per_author_2 = book_counts_per_author_2[book_counts_per_author_2 > 1]

print(multiple_books_per_author_2)

authors
A. Meredith Walters    2
A.A. Milne             6
A.G. Howard            2
A.N. Roquelaure        3
A.S. Byatt             5
                      ..
مصطفى محمود            2
نجيب محفوظ             2
يوسف زيدان             4
村上 春樹                  2
田中メカ]                  2
Name: original_title, Length: 1930, dtype: int64


In [72]:
author_book=books_enriched_cleaned_authors[['goodreads_book_id','original_title', 'authors']]
author_book.head()

goodreads_book_id                            original_title  \
0            2767052                          The Hunger Games   
1                  3  Harry Potter and the Philosopher's Stone   
1                  3  Harry Potter and the Philosopher's Stone   
2              41865                                  Twilight   
3               2657                     To Kill a Mockingbird   

           authors  
0  Suzanne Collins  
1     J.K. Rowling  
1    Mary GrandPré  
2  Stephenie Meyer  
3       Harper Lee

In [73]:
author_book.shape

(13183, 3)

In [74]:
author_book['goodreads_book_id'].dtype

dtype('int64')

In [75]:
author_book['goodreads_book_id'].nunique()

9948

In [76]:
author_book['authors'].dtype

dtype('O')

In [77]:
author_book['authors'].nunique()

6431

In [78]:
# Escape single quotes
author_book['authors'] = author_book['authors'].str.replace("'", "''") # Escape single quotes

# Remove special characters
author_book['authors'] = author_book['authors'] .str.replace("[^\w\s]", "") 

# Remove leading/trailing spaces
author_book['authors'] = author_book['authors'].str.strip() 

In [79]:
author_book['authors'] = author_book['authors'].astype('str')

#Make sure NANs are not strings
author_book['authors'] = author_book['authors'].apply(lambda x: str(x) if pd.notnull(x) else x)

In [80]:
nan_count_authors = author_book['authors'].isnull().sum()
nan_count_authors

0

In [81]:
author_book['original_title'].dtype

dtype('O')

In [82]:
author_book['original_title'].nunique()

9802

In [83]:
# Escape single quotes
author_book['original_title']= author_book['original_title'].str.replace("'", "''") # Escape single quotes

# Remove special characters
author_book['original_title'] = author_book['original_title'] .str.replace("[^\w\s]", "") 

# Remove leading/trailing spaces
author_book['original_title'] = author_book['original_title'].str.strip() 

In [84]:
author_book['original_title']= author_book['original_title'].astype('str')

#Make sure NANs are not strings
author_book['original_title'] = author_book['original_title'].apply(lambda x: str(x) if pd.notnull(x) else x)

In [85]:
nan_count_authors_title = author_book['original_title'].isnull().sum()
nan_count_authors_title

0

In [86]:
print(author_book.isnull().sum())

goodreads_book_id    0
original_title       0
authors              0
dtype: int64


In [87]:
author_book.head()

goodreads_book_id                           original_title          authors
0            2767052                         The Hunger Games  Suzanne Collins
1                  3  Harry Potter and the Philosophers Stone       JK Rowling
1                  3  Harry Potter and the Philosophers Stone    Mary GrandPré
2              41865                                 Twilight  Stephenie Meyer
3               2657                    To Kill a Mockingbird       Harper Lee

In [88]:
author_book.to_csv('New_author_book_SQL.csv', index=False)

#### <font color='purple'> Author table: unique author ID, unique author </font>

In [89]:
# First, let's create a DataFrame of unique authors
authors_df = pd.DataFrame(author_book['authors'].unique(), columns=['author_name'])

# Now, let's add a unique ID for each author.
# We will use the index as ID, but we will add 1 so the IDs start from 1 instead of 0
authors_df['author_id'] = authors_df.index + 1

# Now, each author has a unique ID in the `author_id` column.
print(authors_df)

          author_name  author_id
0     Suzanne Collins          1
1          JK Rowling          2
2       Mary GrandPré          3
3     Stephenie Meyer          4
4          Harper Lee          5
...               ...        ...
5750      Jill Enders       5751
5751    Diane deGroat       5752
5752   Diane de Groat       5753
5753  Steve Pieczenik       5754
5754      John Keegan       5755

[5755 rows x 2 columns]


In [90]:
authors_df.head()

author_name  author_id
0  Suzanne Collins          1
1       JK Rowling          2
2    Mary GrandPré          3
3  Stephenie Meyer          4
4       Harper Lee          5

In [91]:
# Reorder the columns
authors_df = authors_df[['author_id', 'author_name']]
authors_df.head()

author_id      author_name
0          1  Suzanne Collins
1          2       JK Rowling
2          3    Mary GrandPré
3          4  Stephenie Meyer
4          5       Harper Lee

In [92]:
authors_df['author_name'].nunique()

5755

In [93]:
authors_df['author_name'].dtype

dtype('O')

In [94]:
# Escape single quotes
authors_df['author_name']= authors_df['author_name'].str.replace("'", "''") # Escape single quotes

# Remove special characters
authors_df['author_name'] = authors_df['author_name'] .str.replace("[^\w\s]", "") 

# Remove leading/trailing spaces
authors_df['author_name'] = authors_df['author_name'].str.strip() 

In [95]:
authors_df['author_name'] = authors_df['author_name'].astype('str')

#Make sure NANs are not strings
authors_df['author_name'] = authors_df['author_name'].apply(lambda x: str(x) if pd.notnull(x) else x)

In [96]:
nan_count_authors_name= authors_df['author_name'].isnull().sum()
nan_count_authors_name

0

In [97]:
print(authors_df.isnull().sum())

author_id      0
author_name    0
dtype: int64


In [98]:
authors_df['author_id'].dtype

dtype('int64')

In [99]:
authors_df['author_name'] = authors_df['author_name'].astype('str')

In [100]:
def clean_string(s):
    if isinstance(s, str):
        # Replace non-ASCII characters with a single space
        return ''.join(i if ord(i) < 128 else ' ' for i in s)
    else:
        return s

# Apply the cleaning function to all string columns
authors_df = authors_df.applymap(clean_string)


In [101]:
authors_df.head()

author_id      author_name
0          1  Suzanne Collins
1          2       JK Rowling
2          3    Mary GrandPr 
3          4  Stephenie Meyer
4          5       Harper Lee

In [102]:
authors_df.to_csv('authors_df_new.csv', index=False)

#### <font color='purple'> book_auther association table: goodreads_book_id, unique author ID, unique author </font>

In [103]:
#merging the original author_book df with the authors_df to get the author IDs
df_with_author_ids = pd.merge(author_book, authors_df, how='left', left_on='authors', right_on='author_name')

# Creating a new  book_author df
new_book_author = df_with_author_ids[['goodreads_book_id', 'author_id']].copy()

new_book_author.head()

goodreads_book_id  author_id
0            2767052        1.0
1                  3        2.0
2                  3        NaN
3              41865        4.0
4               2657        5.0

In [104]:
#lets do a quick check to see if the values are correct 

ID_to_search = 41865

author_book[author_book['goodreads_book_id'] == ID_to_search]

goodreads_book_id original_title          authors
2              41865       Twilight  Stephenie Meyer

In [105]:
new_book_author.to_csv('new_book_author.csv', index=False)

#### <font color='purple'> genre table:ubique genre id,genres </font>

In [106]:
genres=books_enriched_cleaned[['genres']]
genres.head()

genres
0  (young-adult, fiction, fantasy, science-fictio...
1          (fantasy, fiction, young-adult, classics)
2  (young-adult, fantasy, romance, fiction, paran...
3  (classics, fiction, historical-fiction, young-...
4   (classics, fiction, historical-fiction, romance)

In [107]:
genres["genres"].dtype

dtype('O')

In [108]:
genres["genres"].unique()

array([('young-adult', 'fiction', 'fantasy', 'science-fiction', 'romance'),
       ('fantasy', 'fiction', 'young-adult', 'classics'),
       ('young-adult', 'fantasy', 'romance', 'fiction', 'paranormal'),
       ...,
       ('fiction', 'thriller', 'books', 'mystery', 'suspense', 'science-fiction'),
       ('mystery', 'fiction', 'thriller', 'suspense', 'crime', 'romance', 'contemporary', 'books', 'paranormal'),
       ('history', 'nonfiction', 'historical-fiction')], dtype=object)

In [109]:
# I use explode to get one genre per row
genres = genres.explode('genres')

In [110]:
genres.head(8)

genres
0      young-adult
0          fiction
0          fantasy
0  science-fiction
0          romance
1          fantasy
1          fiction
1      young-adult

In [111]:
# Reset the index
genres = genres.reset_index(drop=True)

In [112]:
#I want to generate a unique id for each genre 

unique_genres = pd.DataFrame(genres['genres'].drop_duplicates())

# Reset the index again to create a unique id for each genre
unique_genres.reset_index(level=0, inplace=True)

# You might want to rename the columns for clarity
unique_genres.columns = ['genre_id', 'genre']

In [113]:
unique_genres.head(20)

genre_id               genre
0          0         young-adult
1          1             fiction
2          2             fantasy
3          3     science-fiction
4          4             romance
5          8            classics
6         13          paranormal
7         16  historical-fiction
8         25        contemporary
9         34             mystery
10        35            thriller
11        36            suspense
12        37               crime
13        52          nonfiction
14        54             history
15        55           biography
16        56              memoir
17       115              poetry
18       126           chick-lit
19       147           christian

In [114]:
def contains_non_ascii(s):
    return any(ord(c) >= 128 for c in s)

# Find rows that contain non-ASCII characters
problem_rows = unique_genres[unique_genres.apply(lambda row: row.astype(str).apply(contains_non_ascii).any(), axis=1)]

print(problem_rows)

Empty DataFrame
Columns: [genre_id, genre]
Index: []


In [115]:
# Function to remove non-ASCII characters from a string
def remove_non_ascii(s):
    if isinstance(s, str):
        return "".join(c for c in s if ord(c) < 128)
    else:
        return s

unique_genres['genre'] = unique_genres['genre'].apply(remove_non_ascii)

In [116]:
problem_rows = unique_genres[unique_genres.apply(lambda row: row.astype(str).str.contains('Ã').any(), axis=1)]
problem_rows

Empty DataFrame
Columns: [genre_id, genre]
Index: []

In [117]:
unique_genres['genre'].dtype

dtype('O')

In [118]:
unique_genres['genre_id'].dtype

dtype('int64')

In [119]:
unique_genres.to_csv('new_unique_genres.csv', index=False)

In [120]:
Final_unique_genres = pd.read_csv('new_unique_genres.csv', encoding='utf-8')
Final_unique_genres.head()

genre_id            genre
0         0      young-adult
1         1          fiction
2         2          fantasy
3         3  science-fiction
4         4          romance

In [121]:
Final_unique_genres.to_csv('Final_unique_genres.csv', index=False)

#### <font color='purple'> book_genre association table:goodreads_book_id, genre Id,genres </font>

#### I want to merge the original books_enriched_cleaned and the unique_genres df but iin the original table I have tuple of genres in each row, so I have to explode the genres column to be able to do the merge later

In [122]:
# Step 1: Explode 'genres' column
books_enriched_cleaned_exploded = books_enriched_cleaned.explode('genres')

books_enriched_cleaned_exploded = books_enriched_cleaned_exploded.reset_index(drop=True)

In [123]:
books_enriched_cleaned_exploded.head(5)

index            authors  average_rating  best_book_id  book_id  \
0      0  [Suzanne Collins]            4.34       2767052        1   
1      0  [Suzanne Collins]            4.34       2767052        1   
2      0  [Suzanne Collins]            4.34       2767052        1   
3      0  [Suzanne Collins]            4.34       2767052        1   
4      0  [Suzanne Collins]            4.34       2767052        1   

   books_count                                        description  \
0          272  WINNING MEANS FAME AND FORTUNE.LOSING MEANS CE...   
1          272  WINNING MEANS FAME AND FORTUNE.LOSING MEANS CE...   
2          272  WINNING MEANS FAME AND FORTUNE.LOSING MEANS CE...   
3          272  WINNING MEANS FAME AND FORTUNE.LOSING MEANS CE...   
4          272  WINNING MEANS FAME AND FORTUNE.LOSING MEANS CE...   

            genres  goodreads_book_id  \
0      young-adult            2767052   
1          fiction            2767052   
2          fantasy            2767052   
3  science-fiction            2767052   
4          romance            2767052   

                                           image_url  ...  ratings_1  \
0  https://images.gr-assets.com/books/1447303603m...  ...      66715   
1  https://images.gr-assets.com/books/1447303603m...  ...      66715   
2  https://images.gr-assets.com/books/1447303603m...  ...      66715   
3  https://images.gr-assets.com/books/1447303603m...  ...      66715   
4  https://images.gr-assets.com/books/1447303603m...  ...      66715   

  ratings_2  ratings_3 ratings_4  ratings_5 ratings_count  \
0    127936     560092   1481305    2706317       4780653   
1    127936     560092   1481305    2706317       4780653   
2    127936     560092   1481305    2706317       4780653   
3    127936     560092   1481305    2706317       4780653   
4    127936     560092   1481305    2706317       4780653   

                                     small_image_url  work_id  \
0  https://images.gr-assets.com/books/1447303603s...  2792775   
1  https://images.gr-assets.com/books/1447303603s...  2792775   
2  https://images.gr-assets.com/books/1447303603s...  2792775   
3  https://images.gr-assets.com/books/1447303603s...  2792775   
4  https://images.gr-assets.com/books/1447303603s...  2792775   

   work_ratings_count  work_text_reviews_count  
0             4942365                   155254  
1             4942365                   155254  
2             4942365                   155254  
3             4942365                   155254  
4             4942365                   155254  

[5 rows x 26 columns]

In [124]:

books_enriched_cleaned_exploded = books_enriched_cleaned_exploded.drop('index', axis=1)

In [125]:
# Merge 'books_enriched_cleaned_exploded' with 'unique_genres'
book_genre_new = pd.merge(books_enriched_cleaned_exploded, unique_genres, how='left', left_on='genres', right_on='genre')

# Keep only the necessary columns
book_genre_new = book_genre_new[['genre_id', 'goodreads_book_id']]

In [126]:
book_genre_new.head()

genre_id  goodreads_book_id
0         0            2767052
1         1            2767052
2         2            2767052
3         3            2767052
4         4            2767052

In [127]:
book_genre_new.shape

(46970, 2)

In [128]:
book_genre_new["genre_id"].nunique()

39

In [129]:
#lets do a quick check to see if the values are correct 

ID_to_search = 2767052

book_genre_new[book_genre_new['goodreads_book_id'] == ID_to_search]

genre_id  goodreads_book_id
0         0            2767052
1         1            2767052
2         2            2767052
3         3            2767052
4         4            2767052

In [130]:
book_genre_new.shape

(46970, 2)

In [131]:
book_genre_new = book_genre_new.drop_duplicates()

In [132]:
book_genre_new.shape

(46970, 2)

In [133]:
book_genre_new.to_csv('book_genre_new.csv', index=False)